rust/basic_bitcoin/src/basic_bitcoin/src/bitcoin_wallet.rs

```diff
@@ -0,0 +1,327 @@
+//! A demo of a very bare-bones bitcoin "wallet".
+//!
+//! The wallet here showcases how bitcoin addresses can be be computed
+//! and how bitcoin transactions can be signed. It is missing several
+//! pieces that any production-grade wallet would have, including:
+//!
+//! * Support for address types that aren't P2PKH.
+//! * Caching spent UTXOs so that they are not reused in future transactions.
+//! * Option to set the fee.
+use crate::{bitcoin_api, ecdsa_api};
+use bitcoin::util::psbt::serialize::Serialize;
+use bitcoin::{
+    blockdata::script::Builder, hashes::Hash, Address, AddressType, OutPoint, Script, SigHashType,
+    Transaction, TxIn, TxOut, Txid,
+};
+use ic_btc_types::{MillisatoshiPerByte, Network, Satoshi, Utxo};
+use ic_cdk::print;
+use sha2::Digest;
+use std::str::FromStr;
+
+const SIG_HASH_TYPE: SigHashType = SigHashType::All;
+
+/// Returns the P2PKH address of this canister at the given derivation path.
+pub async fn get_p2pkh_address(network: Network, derivation_path: Vec<Vec<u8>>) -> String {
+    // Fetch the public key of the given derivation path.
+    let public_key = ecdsa_api::ecdsa_public_key(derivation_path).await;
+
+    // Compute the address.
+    public_key_to_p2pkh_address(network, &public_key)
+}
+
+/// Sends a transaction to the network that transfers the given amount to the
+/// given destination, where the source of the funds is the canister itself
+/// at the given derivation path.
+pub async fn send(
+    network: Network,
+    derivation_path: Vec<Vec<u8>>,
+    dst_address: String,
+    amount: Satoshi,
+) {
+    // Get fee percentiles from previous transactions to estimate our own fee.
+    let fee_percentiles = bitcoin_api::get_current_fee_percentiles(network).await;
+
+    let fee_per_byte = if fee_percentiles.is_empty() {
+        // There are no fee percentiles. This case can only happen on a regtest
+        // network where there are no non-coinbase transactions. In this case,
+        // we use a default of 1000 millisatoshis/byte (i.e. 1 satoshi/byte)
+        1000
+    } else {
+        // Choose the 25th percentile for sending fees to minimize the cost.
+        fee_percentiles[24]
+    };
+
+    // Fetch our public key, P2PKH address, and UTXOs.
+    let own_public_key = ecdsa_api::ecdsa_public_key(derivation_path.clone()).await;
+    let own_address = public_key_to_p2pkh_address(network, &own_public_key);
+
+    print("Fetching UTXOs...");
+    let own_utxos = bitcoin_api::get_utxos(network, own_address.clone())
+        .await
+        .utxos;
+
+    let own_address = Address::from_str(&own_address).unwrap();
+    let dst_address = Address::from_str(&dst_address).unwrap();
```

Comment:
These two lines should be at the beginning of the `send` function in order to avoid falling under this case:

> Unfortunately, this means that if your canister made a call and then panicked in the callback, the canister might never release the resources it allocated for the call.

[Source](https://mmapped.blog/posts/01-effective-rust-canisters.html#async)

Reply:
In this case, there are no resources that we're allocating such that panicking after an `await` would cause trouble.

Panicking after an `await` is unavoidable in our case actually, since we don't know our `own_address` until we make a call to `ecdsa_public_key`, so moving these two lines to the top doesn't really change much.